### DISCUSSION TITLE
A Problem about Inverse Dynamics

### DISCUSSION THREAD

**Discussion by XuSen-Proto** [raisimTech/raisimLib#274] [General]
I found a problem about inverse dynamics recently. I imported iwwa14, a fixed-base manipulator, and set the damping zero, to calculate the NonlinearForce, However, the result turned out different from that solved by robotics system toolbox of MATLAB, which confused me. The code is in the following link.  In the code, I just set the gc, gv, and then call getNonlinearities so the PD controller may not affect the result. I'm looking forward to your reply.
https://github.com/raisimTech/raisimLib/files/8130692/InverseDynamicsTest.zip

**Comment by jhwangbo** [comment]
How do you compute nonlinearities using the MATLAB toolbox? does that include the gravity term?

**Reply by XuSen-Proto** [reply]
I used the robotics system toolbox. I imported the urdf file, gave the gc, i.e. the joint angle, and its derivative, set the gravity, and called the InverseDynamic function.

**Reply by jhwangbo** [reply]
You are comparing different things. Nonlinearity is not inverseDynamics. Please check the documentation here: https://raisim.com/sections/ArticulatedSystem.html#dynamics

**Reply by XuSen-Proto** [reply]
In the code I set the joint acceleration u_dot zero, so τ=h.

**Reply by jhwangbo** [reply]
I cannot run your code. It says "reference to non-existent field 'addCollision'".
I doubt that any of the dynamics methods can be incorrect. They are heavily unit tested against other libraries.
Does switching the order of gc, gv methods help?

**Reply by XuSen-Proto** [reply]
Thanks for your advice. I switched gc and gv,  but it didn't help. The bug you referred occurred because you used an old version of robotics system toolbox, which does not contain addCollision function. In the following attachment, you can replace the iwwa14.m file with iwwa14_nc.m, which might work. Besides, In a recent test, I gave the gc, gv, and zeroed the acceleration to get the joint torque **_τ_**, which equals nonLinearForce. Then I set the GeneralizedForce **_τ_** and integrated for a small time step. In this case, the gv should conserve due to the zero acceleration.  The result shows that **_τ_** calculated from matlab has smaller error. Thus, I think there might be some problem in getNonLinearForce function. The code is in the attachment, you can put it in the previously downloaded folder and have a test.
[IDTestForFixedBaseWithInt.zip](https://github.com/raisimTech/raisimLib/files/8174759/IDTestForFixedBaseWithInt.zip)

**Reply by XuSen-Proto** [reply]
I think I have found the problem. The code is in the attachment. You can see that only after the function "getMassMatrix" is called, the result will equal that solved by matlab. Maybe some variables fail to be updated before "getNonlinearities" is called. I'm looking forward to your reply.
[InverseDynamicsTest.zip](https://github.com/raisimTech/raisimLib/files/8305758/InverseDynamicsTest.zip)
.

**Reply by jhwangbo** [reply]
You are right. All my tests call getMassmatrix first and I didn't notice that error. After v1.1, we don't use mass matrix and nonlinearities for simulation so they are not computed unless you call the methods to compute them. I will think about a better api...
